src/modules/ekf2/EKF/mag_control.cpp

```diff
@@ -41,199 +41,124 @@
-				_saved_mag_bf_variance.zero();
-
-				_control_status.flags.mag_fault = false;
-
-				_mag_lpf.reset(mag_sample.mag);
-				_mag_counter = 1;
-
-			} else {
-				_mag_lpf.update(mag_sample.mag);
-				_mag_counter++;
-			}
-
-			// if enabled, use knowledge of theoretical magnetic field vector to calculate a synthetic magnetomter Z component value.
-			// this is useful if there is a lot of interference on the sensor measurement.
-			if (_params.synthesize_mag_z && (_params.mag_declination_source & GeoDeclinationMask::USE_GEO_DECL)
-			    && (_NED_origin_initialised || PX4_ISFINITE(_mag_declination_gps))
-			   ) {
-				const Vector3f mag_earth_pred = Dcmf(Eulerf(0, -_mag_inclination_gps, _mag_declination_gps)) * Vector3f(_mag_strength_gps, 0, 0);
-				mag_sample.mag(2) = calculate_synthetic_mag_z_measurement(mag_sample.mag, mag_earth_pred);
-				_control_status.flags.synthetic_mag_z = true;
-
-			} else {
-				_control_status.flags.synthetic_mag_z = false;
-			}
-
-			mag_data_valid = checkMagField(mag_sample.mag);
+	// reset the flight alignment flag so that the mag fields will be
+	//  re-initialised next time we achieve flight altitude
+	if (!_control_status_prev.flags.in_air && _control_status.flags.in_air) {
+		_control_status.flags.mag_aligned_in_flight = false;
+	}
 
-			// compute mag heading innovation (for estimator_aid_src_mag_heading logging)
-			const Vector3f mag_observation = mag_sample.mag - _state.mag_B;
-			const Dcmf R_to_earth = updateYawInRotMat(0.f, _R_to_earth);
-			const Vector3f mag_earth_pred = R_to_earth * mag_observation;
+	// check mag state observability
+	checkYawAngleObservability();
+	checkMagBiasObservability();
 
-			resetEstimatorAidStatus(_aid_src_mag_heading);
-			_aid_src_mag_heading.timestamp_sample = mag_sample.time_us;
-			_aid_src_mag_heading.observation = -atan2f(mag_earth_pred(1), mag_earth_pred(0)) + getMagDeclination();
-			_aid_src_mag_heading.innovation = wrap_pi(getEulerYaw(_R_to_earth) - _aid_src_mag_heading.observation);
+	if (_mag_bias_observable || _yaw_angle_observable) {
+		_time_last_mov_3d_mag_suitable = _time_delayed_us;
+	}
 
-			// compute magnetometer innovations (for estimator_aid_src_mag logging)
-			//  rotate magnetometer earth field state into body frame
-			const Vector3f mag_I_body = _state.quat_nominal.rotateVectorInverse(_state.mag_I);
-			const Vector3f mag_innov = mag_I_body - mag_observation;
+	if (_params.mag_fusion_type == MagFuseType::NONE) {
+		stopMagFusion();
+		stopMagHdgFusion();
+		return;
+	}
 
-			resetEstimatorAidStatus(_aid_src_mag);
-			_aid_src_mag.timestamp_sample = mag_sample.time_us;
-			mag_observation.copyTo(_aid_src_mag.observation);
-			mag_innov.copyTo(_aid_src_mag.innovation);
+	// stop mag (require a reset before using again) if there was an external yaw reset (yaw estimator, GPS yaw, etc) or yaw alignment change
+	if (_mag_decl_cov_reset) {
+		const bool yaw_estimator_reset = _information_events.flags.yaw_aligned_to_imu_gps;
```

Comment:
If there is a yaw estimator reset, the quat reset counter is also incremented so this condition is unnecessary.